apps/transport/test/transport_web/controllers/backoffice/notification_subscription_controller_test.exs

```diff
@@ -8,59 +8,130 @@ defmodule TransportWeb.NotificationSubscriptionControllerTest do
-             },
-             %DB.NotificationSubscription{
-               contact_id: ^contact_id,
-               dataset_id: ^dataset_id,
-               source: :admin,
-               reason: :resource_unavailable
-             }
-           ] = DB.NotificationSubscription |> DB.Repo.all() |> Enum.sort_by(& &1.reason)
-
-    # Sending a request again, but with a new valid reason. Should not try to create duplicates.
-    conn_response =
-      conn
-      |> setup_admin_in_session()
-      |> post(backoffice_notification_subscription_path(conn, :create, Map.put(args, "dataset_with_error", "true")))
+  describe "create" do
+    test "for reasons related to a dataset", %{conn: conn} do
+      %DB.Dataset{id: dataset_id} = insert(:dataset)
+      %DB.Contact{id: contact_id} = DB.Contact.insert!(sample_contact_args())
+
+      args = %{
+        "redirect_location" => "dataset",
+        "dataset_id" => dataset_id,
+        "contact_id" => contact_id,
+        "expiration" => "true",
+        "resource_unavailable" => "true",
+        # Should be ignored because this is not a valid reason
+        "ignored_notification_reason" => "true",
+        # Ignored because it's set to false (happens only when manipulating requests)
+        "dataset_with_error" => "false"
+      }
+
+      conn_response =
+        conn
+        |> setup_admin_in_session()
+        |> post(backoffice_notification_subscription_path(conn, :create, args))
+
+      assert redirected_to(conn_response, 302) == backoffice_page_path(conn, :edit, dataset_id) <> @target_html_anchor
+      assert get_flash(conn_response, :info) =~ "L'abonnement a été créé"
+
+      assert [
+               %DB.NotificationSubscription{
+                 contact_id: ^contact_id,
+                 dataset_id: ^dataset_id,
+                 source: :admin,
+                 reason: :expiration
+               },
+               %DB.NotificationSubscription{
+                 contact_id: ^contact_id,
+                 dataset_id: ^dataset_id,
+                 source: :admin,
+                 reason: :resource_unavailable
+               }
+             ] = DB.NotificationSubscription |> DB.Repo.all() |> Enum.sort_by(& &1.reason)
+
+      # Sending a request again, but with a new valid reason. Should not try to create duplicates.
+      conn_response =
+        conn
+        |> setup_admin_in_session()
+        |> post(backoffice_notification_subscription_path(conn, :create, Map.put(args, "dataset_with_error", "true")))
+
+      assert redirected_to(conn_response, 302) == backoffice_page_path(conn, :edit, dataset_id) <> @target_html_anchor
+      assert get_flash(conn_response, :info) =~ "L'abonnement a été créé"
+
+      assert MapSet.new([:dataset_with_error, :expiration, :resource_unavailable]) ==
+               DB.NotificationSubscription.base_query()
+               |> select([notification_subscription: ns], ns.reason)
+               |> DB.Repo.all()
+               |> MapSet.new()
+    end
```

Comment:
Ce test était déjà là. Il ne change pas, il est juste mis dans un `describe`

Reply:
Il contient un souci je crois. "Should not try to create duplicates" n'est en fait pas bien testé, à cause de l'utilisation d'un `MapSet` qui dédoublonne:

```
iex(2)> MapSet.new([:dataset_with_error, :expiration, :resource_unavailable, :expiration]) 
MapSet.new([:dataset_with_error, :expiration, :resource_unavailable])
```

Du coup si ça crée des doublons, on ne le verra pas !

Reply:
https://github.com/etalab/transport-site/pull/3004/commits/63deee0f155afa167aaafbf0b5d24d0636a88ddd en plus, t'as raison faut mieux être prudent !